src/interrupts/mod.rs

```diff
@@ -0,0 +1,577 @@
+    /// });
+    /// ```
+    pub fn register<F>(&mut self,
+                              irq: InterruptRequest,
+                              priority: Priority,
+                              mut isr: F)
+                              -> Result<InterruptHandle<()>, Error>
+        where F: FnMut() + 'a + Send
+    {
+        self.register_owned(irq, priority, (), move |_| {isr()})
+    }
+
+    /// Registers an interrupt with the lifetime of the `InterruptTable` and pass ownership of a variable `owned_data: T` that is passed
+    /// to the `isr(&mut T)` when the corresponding interupt `irq` occur.
+    /// 
+    /// The ownership of the data `owned_data` is returned in the `unregister()` function
+    ///
+    /// # Examples
+    /// ```
+    /// use stm32f7::interrupts::interrupt_request::InterruptRequest::Tim7;
+    /// use interrupts::Priority::P1;    
+    /// interrupts::scope(nvic, |irq| { hprintln!("Default handler: {}", irq) },
+    ///     |interrupt_table| {
+    ///             let data: SomeData = ...;
+    ///             let interrupt_handle = interrupt_table.register_owned(Tim7, P1, data,
+    ///             |owned_data| {
+    ///                 // Isr for interrupt `Tim7`
+    ///                 owned_data.do_anything();
+    ///             }).expect("Interrupt already used");
+    ///
+    ///             /* Code that needs interrupt `Tim7` */
+    ///
+    ///             // Unregister interrupt and get back the ownership to `data`
+    ///             let data = interrupt_table.unregister(interrupt_handle).unwrap();
+    /// });
+    /// ```
+    pub fn register_owned<F, T>(&mut self,
+                                 irq: InterruptRequest,
+                                 priority: Priority,
+                                 owned_data: T,
+                                 mut isr: F)
+                                 -> Result<InterruptHandle<T>, Error>
+        where   T: Send,
+                F: FnMut(&mut T) + 'a + Send
+    {
+        if unsafe{ISRS[irq as usize].is_some()} {
+            return Err(Error::InterruptAlreadyInUse(irq));
+        }
+        // Insert data only, when interrupt isn't used, therefore nobody reads the data => no dataraces
+        self.data[irq as usize] = unsafe {
+            transmute::<Option<*mut T>, Option<*mut ()>>(Some(Box::into_raw(Box::new(owned_data))))
+        };
+        
+        // transmute::<Box<FnMut()>, Box<FnMut() + 'static + Send>> is safe, because of the drop implementation of InterruptTable ('static is not needed for closure)
+        // and alway only one isr can access the data (Send is not needed for closure)
+        let isr = unsafe {
+            transmute::<Box<FnMut()>, Box<FnMut() + 'static + Send>>(Box::new(
+            || {
+                match self.data[irq as usize] {
+                    // Safe, since the correct type is known
+                    Some(ptr) => isr(Box::from_raw(transmute::<*mut (), *mut T>(ptr)).as_mut()),
```

Comment:
I think this must be written as `isr(*(ptr as *mut T))`. Otherwise the `Box` is freed after the first time the interrupt is called. You just want a reference here.